File: ai_wealth_sdk/lib/src/predictive/domain/entities/prediction.dart
import 'package:equatable/equatable.dart';

enum PredictionType {
  salaryCredit,
  billDue,
  emiDue,
  lowBalance,
  taxReminder;

  static PredictionType fromWire(String value) => switch (value) {
        'salary_credit' => PredictionType.salaryCredit,
        'bill_due' => PredictionType.billDue,
        'emi_due' => PredictionType.emiDue,
        'low_balance' => PredictionType.lowBalance,
        'tax_reminder' => PredictionType.taxReminder,
        _ => PredictionType.billDue,
      };
}

enum PredictionSeverity {
  info,
  warning,
  critical;

  static PredictionSeverity fromWire(String value) => switch (value) {
        'critical' => PredictionSeverity.critical,
        'warning' => PredictionSeverity.warning,
        _ => PredictionSeverity.info,
      };
}

/// A single forward-looking prediction.
class Prediction extends Equatable {
  const Prediction({
    required this.type,
    required this.title,
    required this.message,
    required this.predictedDate,
    required this.daysAway,
    required this.severity,
    this.amount,
  });

  final PredictionType type;
  final String title;
  final String message;
  final DateTime predictedDate;
  final int daysAway;
  final PredictionSeverity severity;
  final double? amount;

  @override
  List<Object?> get props =>
      [type, title, message, predictedDate, daysAway, severity, amount];
}

File: backend/app/models/prediction.py
"""Predictive Banking domain models.

Predictions are computed on the fly by the prediction engine, not persisted.
`RecurringItem` is the intermediate the engine projects forward.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from enum import Enum

from app.models.spending import SpendCategory, TransactionDirection


class PredictionType(str, Enum):
    salary_credit = "salary_credit"
    bill_due = "bill_due"
    emi_due = "emi_due"
    low_balance = "low_balance"
    tax_reminder = "tax_reminder"


class PredictionSeverity(str, Enum):
    info = "info"
    warning = "warning"
    critical = "critical"


@dataclass
class RecurringItem:
    """A detected recurring cashflow (salary, rent, EMI, subscription…)."""

    label: str
    amount: float
    direction: TransactionDirection
    day: int  # day-of-month it typically occurs
    category: SpendCategory
    is_loan: bool = False


@dataclass
class Prediction:
    type: PredictionType
    title: str
    message: str
    predicted_date: date
    severity: PredictionSeverity
    days_away: int
    amount: float | None = None
